### PR TITLE
[RFC] Upstream Versioning

### PR DESCRIPTION
**Pull request by wurbanski**

### CODE REVIEW

File: source/rfc/versioning.rst [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Problem
+-------
+
+Upstream versions of operating systems and other software (e.g. OpenStack) change in schedules different than that of Contrail's. Red Hat family of operating systems replaces a previous version with new one and drops the support for older ones. There are multiple combinations of supported OpenStack and CentOS/Red Hat releases and they all have to be tracked.
+Using official repositories makes CI setup prone to network failures.
+    
+Solution
+--------
+
+Life cycle for external/upstream components has to be tracked in order to accomodate for changes in operating systems' and other software's versioning.
```
Comment (kravvcuCL) [inline source/rfc/versioning.rst:18]: `Lifecycle of`
IMO, not sure about the space tho

File: source/rfc/versioning.rst [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Additionally, any packages or source code that is used in CI should be mirrored for use in air-gapped environment to reduce dependencies on external sources and proneness to network failures.
+
+
+Tracking release timelines
+-----------------------------
```
Comment (kravvcuCL) [inline source/rfc/versioning.rst]: Should not be the same length as the title?

**Review by kravvcuCL — CHANGES_REQUESTED** [review]

**Review by kravvcuCL — APPROVED** [review]

File: source/rfc/versioning.rst (hunk 1 above)
Comment (pjrusak) [reply source/rfc/versioning.rst:18]: Looks good to me

File: source/rfc/versioning.rst [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+Using official repositories makes CI setup prone to network failures.
+    
+Solution
+--------
+
+Life cycle for external/upstream components has to be tracked in order to accomodate for changes in operating systems' and other software's versioning.
+Additionally, any packages or source code that is used in CI should be mirrored for use in air-gapped environment to reduce dependencies on external sources and proneness to network failures.
+
+Tracking release timelines
+-----------------------------
```
Comment (kravvcuCL) [inline source/rfc/versioning.rst]: Make equal length to header

File: source/rfc/versioning.rst [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+2. **CentOS**
+
+   *Note: CentOS Release schedule follows that of Red Hat's closely.*
+
+   * `About CentOS <https://wiki.centos.org/About/Product>`_
+
+3. **Ubuntu**
+
+   * `About Ubuntu 16.04 <http://releases.ubuntu.com/16.04/>`_
```
Comment (kravvcuCL) [inline source/rfc/versioning.rst]: https://wiki.ubuntu.com/Releases

File: source/rfc/versioning.rst [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+   Expected structure of a mirrored repository::
+
+     7.0.1406
+     7.1.1503
+     7.2.1511
+     7.3.1611
+     7.4.1708
+     7.5.1804
+     7 -> 7.5.1804
+      \
```
Comment (kravvcuCL) [inline source/rfc/versioning.rst]: not exactly sure what does this mean?

**Review by kravvcuCL — CHANGES_REQUESTED** [review]

**Review by kravvcuCL — DISMISSED** [review]

File: source/rfc/versioning.rst [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+^^^^^^^^^
+
+All information about OpenStack releases can be found on `OpenStack Releases website <https://releases.openstack.org>`_.
+
+Red Hat OpenStack Platform
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+All information can be found on `Red Hat's Website <https://access.redhat.com/support/policy/updates/openstack/platform>`_.
+
+Handling of upcoming versions
```
Comment (jluk-codi) [inline source/rfc/versioning.rst:52]: Describe or link to info about:
- how to use RHEL beta repos
- how to use CentOS CR repos
- how to get kolla containers for unreleased openstack (master builds?)

File: source/rfc/versioning.rst [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+For CentOS it is available as a CR (`Continuous Release <https://wiki.centos.org/AdditionalResources/Repositories/CR>`_) repository.
+
+For Red Hat repositories with suffix `-beta-rpms` can be used, e.g. `rhel-7-server-beta-rpms`. No additional license required.
+
+Targeting specific releases
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+*Note: This section relates to CentOS only.*
+
+To make sure that updates do not happen without operation team's knowledge, current version symlink (7) should be set for currently used point-release manually, e.g. `7 -> 7.5.1804` at the time of writing.
```
Comment (jluk-codi) [inline source/rfc/versioning.rst]: Kinda very specific implementation detail?

File: source/rfc/versioning.rst [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+Upstream versions of operating systems and other software (e.g. OpenStack) change in schedules different than that of Contrail's. Red Hat family of operating systems replaces a previous version with new one and drops the support for older ones. There are multiple combinations of supported OpenStack and CentOS/Red Hat releases and they all have to be tracked.
+Using official repositories makes CI setup prone to network failures.
+    
+Solution
+--------
+
+Life cycle for external/upstream components has to be tracked in order to accomodate for changes in operating systems' and other software's versioning.
+Additionally, any packages or source code that is used in CI should be mirrored for use in air-gapped environment to reduce dependencies on external sources and proneness to network failures.
+
+Tracking release timelines
```
Comment (jluk-codi) [inline source/rfc/versioning.rst:21]: You can mention which distributions of those project are used (packages, docker images, cloud images etc.). They can have specific release plans.

File: source/rfc/versioning.rst (hunk 6 above)
Comment (wurbanski) [reply source/rfc/versioning.rst:52]: I think that both RHEL and CentOS is described below.
For CentOS, use CR repo, for redhat `*-beta-rpms`

File: source/rfc/versioning.rst (hunk 7 above)
Comment (wurbanski) [reply source/rfc/versioning.rst]: Partially agree - it should be re-worded to hide implementation but information should be kept.

File: source/rfc/versioning.rst (hunk 8 above)
Comment (wurbanski) [reply source/rfc/versioning.rst:21]: Will do.

**Review by kravvcuCL — DISMISSED** [review]

File: source/rfc/versioning.rst (hunk 6 above)
Comment (jluk-codi) [reply source/rfc/versioning.rst:52]: This still needs more details, the beta repo doesn't have all the packages.

File: source/rfc/versioning.rst [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+*Note: This section relates to CentOS only.*
+
+To make sure that updates do not happen without operation team's knowledge, current version symlink (7) should be set for currently used point-release manually, e.g. `7 -> 7.5.1804` at the time of writing.
+
+OpenStack
+^^^^^^^^^
+
+OpenStack repositories are currently being used for CentOS only. They might be provided ahead of time if there is a `SIG <https://wiki.centos.org/SpecialInterestGroup>`_ working on packages for that release. The packages are available in `cloud` repositories for CentOS and maintained by `Cloud SIG <https://wiki.centos.org/SpecialInterestGroup/Cloud>`_.
```
Comment (jluk-codi) [inline source/rfc/versioning.rst:88]: There are master builds for RDO project available, they can be used to test newer versions in advance. A lot of info here: https://www.rdoproject.org/what/

**Review by jluk-codi — APPROVED** [review]

**Pull request merged**